File: ios/Classes/ProfileManager.swift
import Foundation

/// Persists blocking profiles in `UserDefaults` and activates/deactivates them
/// by delegating to `ShieldManager`.
///
/// A profile groups a set of app identifiers (and optionally schedules) that
/// can be blocked atomically. Activating a profile deactivates any previously
/// active profile first. The active profile ID is stored separately from the
/// profile list so that `getProfiles` can annotate each profile with its
/// current active state without mutating the stored list.
class ProfileManager: NSObject {

    private let userDefaultsKey = "app_blocker_profiles"
    private let activeProfileKey = "app_blocker_active_profile_id"

    // Weak reference so ShieldManager's lifetime is not prolonged by ProfileManager.
    private weak var shieldManager: ShieldManager?

    init(shieldManager: ShieldManager) {
        self.shieldManager = shieldManager
        super.init()
    }

    // MARK: - CRUD

    /// Adds [data] as a new profile, replacing any existing profile with the
    /// same `"id"` key. New profiles default to `isActive = false`.
    func createProfile(data: [String: Any]) {
        guard let id = data["id"] as? String else { return }
        var profiles = loadProfiles()
        profiles.removeAll { ($0["id"] as? String) == id }
        var profileData = data
        if profileData["isActive"] == nil { profileData["isActive"] = false }
        profiles.append(profileData)
        saveProfiles(profiles)
    }

    /// Replaces the profile with the same `"id"`, or appends it if absent.
    func updateProfile(data: [String: Any]) {
        guard let id = data["id"] as? String else { return }
        var profiles = loadProfiles()
        if let index = profiles.firstIndex(where: { ($0["id"] as? String) == id }) {
            profiles[index] = data
        } else {
            profiles.append(data)
        }
        saveProfiles(profiles)
    }

    /// Deletes the profile with [id], deactivating it first if it is active.
    func deleteProfile(id: String) {
        let activeId = UserDefaults.standard.string(forKey: activeProfileKey)
        if activeId == id {
            shieldManager?.unblockAll()
            UserDefaults.standard.removeObject(forKey: activeProfileKey)
        }
        var profiles = loadProfiles()
        profiles.removeAll { ($0["id"] as? String) == id }
        saveProfiles(profiles)
    }

    /// Returns all profiles with an injected `"isActive"` field reflecting the
    /// currently active profile ID.
    func getProfiles() -> [[String: Any]] {
        let activeId = UserDefaults.standard.string(forKey: activeProfileKey)
        return loadProfiles().map { profile in
            var p = profile
            p["isActive"] = (p["id"] as? String) == activeId
            return p
        }
    }

    // MARK: - Activation

    /// Activates the profile with [id], deactivating any previously active profile.
    ///
    /// - Returns: `true` if the profile was found and activated; `false` otherwise.
    @discardableResult
    func activateProfile(id: String) -> Bool {
        let profiles = loadProfiles()
        guard let profile = profiles.first(where: { ($0["id"] as? String) == id }) else {
            return false
        }

        // Deactivate the current profile if it differs.
        let currentActiveId = UserDefaults.standard.string(forKey: activeProfileKey)
        if let currentId = currentActiveId, currentId != id {
            deactivateProfile(id: currentId)
        }

        if let appIdentifiers = profile["appIdentifiers"] as? [String] {
            shieldManager?.blockApps(identifiers: appIdentifiers)
        }

        UserDefaults.standard.set(id, forKey: activeProfileKey)
        return true
    }

    /// Deactivates the profile with [id]. No-op if it is not currently active.
    func deactivateProfile(id: String) {
        guard UserDefaults.standard.string(forKey: activeProfileKey) == id else { return }
        shieldManager?.unblockAll()
        UserDefaults.standard.removeObject(forKey: activeProfileKey)
    }

    /// Returns the active profile with an injected `"isActive": true` field,
    /// or `nil` if no profile is active. Cleans up stale active IDs.
    func getActiveProfile() -> [String: Any]? {
        guard let activeId = UserDefaults.standard.string(forKey: activeProfileKey) else {
            return nil
        }
        guard var profile = loadProfiles().first(where: { ($0["id"] as? String) == activeId }) else {
            // The active profile was deleted without going through deactivateProfile.
            UserDefaults.standard.removeObject(forKey: activeProfileKey)
            return nil
        }
        profile["isActive"] = true
        return profile
    }

    // MARK: - Persistence

    private func loadProfiles() -> [[String: Any]] {
        guard let data = UserDefaults.standard.data(forKey: userDefaultsKey) else {
            return []
        }
        do {
            let decoded = try JSONSerialization.jsonObject(with: data)
            return decoded as? [[String: Any]] ?? []
        } catch {
            // Persisted data is corrupt; reset to a clean state.
            UserDefaults.standard.removeObject(forKey: userDefaultsKey)
            return []
        }
    }

    private func saveProfiles(_ profiles: [[String: Any]]) {
        do {
            let data = try JSONSerialization.data(withJSONObject: profiles)
            UserDefaults.standard.set(data, forKey: userDefaultsKey)
        } catch {
            assertionFailure("Failed to serialise profiles: \(error)")
        }
    }
}

File: android/src/main/kotlin/com/khanhtq/app_blocker/scheduling/ProfileManager.kt
package com.khanhtq.app_blocker.scheduling

import android.content.Context
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.google.gson.reflect.TypeToken
import com.khanhtq.app_blocker.blocking.BlockingServiceManager
import com.khanhtq.app_blocker.event.BlockEventStreamHandler
import com.khanhtq.app_blocker.persistence.BlockerPreferences

/**
 * Represents a blocking profile — a named group of apps and optional schedules
 * that can be activated/deactivated atomically.
 */
data class ProfileData(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String,
    @SerializedName("appIdentifiers") val appIdentifiers: List<String>,
    @SerializedName("schedules") val schedules: List<ScheduleData>,
    @SerializedName("isActive") val isActive: Boolean,
) {
    /** Converts this instance to the map format sent over the Flutter channel. */
    fun toMap(): Map<String, Any?> = mapOf(
        "id" to id,
        "name" to name,
        "appIdentifiers" to appIdentifiers,
        "schedules" to schedules.map { it.toMap() },
        "isActive" to isActive,
    )

    companion object {
        /**
         * Constructs a [ProfileData] from a Flutter channel map.
         *
         * List elements arrive as [List<*>]; [filterIsInstance] recovers type
         * safety without unchecked casts.
         */
        fun fromMap(data: Map<String, Any?>): ProfileData {
            val appIdentifiers = (data["appIdentifiers"] as? List<*>)
                ?.filterIsInstance<String>()
                ?: emptyList()

            val schedules = (data["schedules"] as? List<*>)
                ?.filterIsInstance<Map<String, Any?>>()
                ?.map { ScheduleData.fromMap(it) }
                ?: emptyList()

            return ProfileData(
                id = data["id"] as String,
                name = data["name"] as? String ?: "",
                appIdentifiers = appIdentifiers,
                schedules = schedules,
                isActive = data["isActive"] as? Boolean ?: false,
            )
        }
    }
}

/**
 * Creates, updates, deletes, and activates/deactivates blocking profiles.
 *
 * Profiles are persisted as a JSON array in [BlockerPreferences]. Activating
 * a profile:
 * 1. Deactivates any currently active profile.
 * 2. Starts blocking the profile's apps via [BlockingServiceManager].
 * 3. Registers each of the profile's nested schedules via [ScheduleManager].
 *
 * Deactivating reverses steps 2–3.
 */
class ProfileManager(private val context: Context) {

    companion object {
        private const val PREFS_KEY_PROFILES = "profiles"
    }

    private val preferences = BlockerPreferences(context)
    private val gson = Gson()
    private val scheduleManager = ScheduleManager(context)
    private val blockingServiceManager = BlockingServiceManager(context)

    // ------------------------------------------------------------------
    // CRUD
    // ------------------------------------------------------------------

    /** Persists [data] as a new profile. */
    fun createProfile(data: Map<String, Any?>) {
        val profile = ProfileData.fromMap(data)
        val profiles = loadProfiles().toMutableList()
        profiles.add(profile)
        saveProfiles(profiles)
    }

    /** Replaces the profile with the same id. */
    fun updateProfile(data: Map<String, Any?>) {
        val updated = ProfileData.fromMap(data)
        val profiles = loadProfiles().toMutableList()
        val index = profiles.indexOfFirst { it.id == updated.id }
        if (index < 0) return

        profiles[index] = updated
        saveProfiles(profiles)
    }

    /** Deletes the profile with [id], deactivating it first if necessary. */
    fun deleteProfile(id: String) {
        val profiles = loadProfiles().toMutableList()
        val profile = profiles.find { it.id == id } ?: return

        if (profile.isActive) deactivateInternal(profile)
        profiles.removeAll { it.id == id }
        saveProfiles(profiles)
    }

    /** Returns all profiles as wire-format maps. */
    fun getProfiles(): List<Map<String, Any?>> = loadProfiles().map { it.toMap() }

    // ------------------------------------------------------------------
    // Activation
    // ------------------------------------------------------------------

    /**
     * Activates the profile with [id].
     *
     * @throws IllegalArgumentException if no profile with [id] exists.
     */
    fun activateProfile(id: String) {
        val profiles = loadProfiles().toMutableList()

        // Deactivate any currently active profile.
        val activeIndex = profiles.indexOfFirst { it.isActive }
        if (activeIndex >= 0) {
            deactivateInternal(profiles[activeIndex])
            profiles[activeIndex] = profiles[activeIndex].copy(isActive = false)
        }

        val targetIndex = profiles.indexOfFirst { it.id == id }
        if (targetIndex < 0) throw IllegalArgumentException("Profile '$id' not found.")

        val target = profiles[targetIndex].copy(isActive = true)
        profiles[targetIndex] = target
        saveProfiles(profiles)

        // Block the profile's apps.
        blockingServiceManager.startBlocking(target.appIdentifiers)

        // Register each embedded schedule.
        for (schedule in target.schedules) {
            scheduleManager.addSchedule(schedule.copy(enabled = true).toMap())
        }

        BlockEventStreamHandler.sendEvent(
            mapOf(
                "type" to "profileActivated",
                "profileId" to id,
                "timestamp" to System.currentTimeMillis(),
            )
        )
    }

    /** Deactivates the currently active profile. Returns `true` if one was active. */
    fun deactivateActiveProfile(): Boolean {
        val activeId = loadProfiles().find { it.isActive }?.id ?: return false
        deactivateProfile(activeId)
        return true
    }

    /** Deactivates the profile with [id]. No-op if it is not active. */
    fun deactivateProfile(id: String) {
        val profiles = loadProfiles().toMutableList()
        val index = profiles.indexOfFirst { it.id == id }
        if (index < 0) return

        deactivateInternal(profiles[index])
        profiles[index] = profiles[index].copy(isActive = false)
        saveProfiles(profiles)
    }

    /** Returns the active profile as a wire-format map, or `null`. */
    fun getActiveProfile(): Map<String, Any?>? =
        loadProfiles().find { it.isActive }?.toMap()

    // ------------------------------------------------------------------
    // Private helpers
    // ------------------------------------------------------------------

    /**
     * Stops blocking the [profile]'s apps and removes its schedules.
     * Does **not** write to storage — callers handle persistence.
     */
    private fun deactivateInternal(profile: ProfileData) {
        // Emit individual unblocked events for each app before stopping blocking
        val timestamp = System.currentTimeMillis()
        for (packageName in profile.appIdentifiers) {
            BlockEventStreamHandler.sendEvent(
                mapOf(
                    "type" to "unblocked",
                    "packageName" to packageName,
                    "timestamp" to timestamp,
                )
            )
        }
        
        blockingServiceManager.stopBlockingApps(profile.appIdentifiers)
        for (schedule in profile.schedules) {
            scheduleManager.removeSchedule(schedule.id)
        }
        BlockEventStreamHandler.sendEvent(
            mapOf(
                "type" to "profileDeactivated",
                "profileId" to profile.id,
                "timestamp" to timestamp,
            )
        )
    }

    private fun loadProfiles(): List<ProfileData> {
        val json = preferences.getString(PREFS_KEY_PROFILES, null) ?: return emptyList()
        val type = object : TypeToken<List<ProfileData>>() {}.type
        return try {
            gson.fromJson(json, type) ?: emptyList()
        } catch (_: Exception) {
            emptyList()
        }
    }

    private fun saveProfiles(profiles: List<ProfileData>) {
        preferences.putString(PREFS_KEY_PROFILES, gson.toJson(profiles))
    }
}
